File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-3.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        start, end = 0, 0
        sub_string = set()
        maxLen = 0

        while end < len(s):
            while start <= end and s[end] in sub_string:
                sub_string.remove(s[start])
                start += 1 
            sub_string.add(s[end]) 
            maxLen = max(maxLen, len(sub_string)) 
            end += 1   
        return maxLen

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-3.java
class Solution {
    public int characterReplacement(String s, int k) {
        char[] charS = s.toCharArray();
        Map<Character, Integer> map = new HashMap<>();

        int start = 0;

        int maxFeq = 0;
        int maxLen = 0;
        
        for (int end=0; end < charS.length; end++){
            map.put(charS[end], map.getOrDefault(charS[end], 0) + 1);
            maxFeq = Math.max(maxFeq, map.get(charS[end]));

            while (((end-start+1)-maxFeq) > k) {
                map.put(charS[start], map.get(charS[start]) -1);
                start++;
            }
            maxLen = Math.max(maxLen, end-start+1);
        }

        return maxLen;
    }
}
